### DISCUSSION TITLE
Many-to-Many without table in the other end

### DISCUSSION THREAD

**Discussion by mtvx** [sqlalchemy/sqlalchemy#12109] [Usage Questions] [answered]
Consider a typical Many-to-Many mapping, e.g. Parent-Child [as shown in the examples](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#many-to-many).

In my case, the `Parent` table (and so the model, unless SQLAlchemy support some sort of "virtual models", ie. model without real table) won't actually exist, ie. parent will be just a reference (e.g UUID) in another system. Probably a typical scenario in distributed architectures.

How can this be configured, so that I can still benefit from many-to-many ~magic~ abstraction? E.g. ability to load up all parent IDs for given Child easily (like `child.parents`, which would be a list of UUIDs, instead).

The docs also explain about the [association object](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#association-object), which is more low level way for the same, but also that seems to assume that tables/models at both ends exist.

**Comment by zzzeek** [comment, marked as answer]
this is the "list of scalar values" pattern, you map the association table and if you want to retrieve the list as just that one value rather than an object you add in [association proxy](https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html#simplifying-scalar-collections)

**Reply by mtvx** [reply]
Hmm, I thought that wouldn't fit the use case, but I guess I judged too quickly. Thanks for the pointer, I'll have a look.

**Reply by mtvx** [reply]
I managed to make it work with association object (so, with a model class, not a `Table` instance), something like instructed in https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#combining-association-object-with-many-to-many-access-patterns, e.g.:

```py
class Association(Base):
    ...
    left_id: Mapped[int] = mapped_column(primary_key=True)
    right_id: Mapped[int] = mapped_column(
        ForeignKey("right_table.id"), primary_key=True
    )
    child: Mapped["Child"] = relationship(back_populates="parent_id_associations")


class Child(Base):
    ...
    parent_id_associations: Mapped[list[Association]] = relationship(back_populates="child")
    parent_ids: AssociationProxy[list[int]] = association_proxy(
        "parent_id_associations",
        "left_id",
        creator=lambda parent_id: Association(left_id=parent_id),
    )
```

This seems to work.

But is it possible to use `AssociationProxy` directly with `Table` and `secondary` relationships, ie. the basic https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#setting-bi-directional-many-to-many form like:

```py
association_table = Table(
    "association_table",
    Base.metadata,
    Column("left_id", primary_key=True),
    Column("right_id", ForeignKey("right_table.id"), primary_key=True),
)


class Child(Base):
    parent_ids: Mapped[list[int]]  # <-- how?
```

I'm fine using the first form, just curious if the second form can be achieved, too?

**Reply by zzzeek** [reply]
I thought you wanted a `list[UUID]`, you wrote `list[int]`.       `Mapped[list[UUID]]` and `AssociationProxy[list[UUID]]` are both instances of `ORMDescriptor[list[UUID]]` if you prefer typing it that way.       other than that I dont see what's different with your two `Child` examples unless you are trying to have less configurational code

**Reply by mtvx** [reply]
Sorry, my DB has `UUID` columns, but the docs examples have `int` so I converted to it but forgot to change that one type (converted now in the above reply). Doesn't really matter which one it is.

> unless you are trying to have less configurational code

Yes, just wondering how to do the same with `Table` syntax. This bit in the second snippet:

```py
class Child(Base):
    parent_ids: Mapped[list[int]]  # <-- how?
```

**Reply by zzzeek** [reply]
unless you want to use the PostgreSQL ARRAY datatype or something like that, in relational databases you can only have collections of rows in other tables, not collections of scalar objects.   so at the mapping level a `list[]` can only be against another mapped class.

**Reply by mtvx** [reply]
Yeah, I won't be adjusting DDL (it's an existing database) so going `ARRAY` is a no go.

If the first form works, and the DDL is the same, I guess there's no technical restriction in supporting the second form (ie. with `Table`)? Maybe SQLAlchemy just doesn't support it (and maybe it doesn't need to), but technically it would be doable? (If not, I'm just failing to understand the topic, and we can leave it to a rest 😄)

---

> you can only have collections of rows in other tables

The association table can exist without a foreign key to the other table. For the sake of example, let's say it's a User-Car many-to-many mapping (where users may own/use multiple cars, and cars may have multiple owners).

Now consider a "micro-service" (well, distributed) architecture, where the user identity (and details) exists in a authentication/identity system and the car data exists another system. In this case, relationship to the user is managed with a simple "user_id" column (not a foreign key) and any user details are fetched on-the-fly, where needed. This is the scenario I'm mapping (or rather, converting an old monolithic system/DB into).

And yeah, it seems to work fine in the "association object" model. But as it's a common use case, maybe a more streamlined syntax could be nice...(?)

**Reply by zzzeek** [reply]
> Yeah, I won't be adjusting DDL (it's an existing database) so going `ARRAY` is a no go.
> 
> If the first form works, and the DDL is the same, I guess there's no technical restriction in supporting the second form (ie. with `Table`)? Maybe SQLAlchemy just doesn't support it (and maybe it doesn't need to), but technically it would be doable? (If not, I'm just failing to understand the topic, and we can leave it to a rest 😄)

it's not technically doable at all, the `Mapped` construct means a specific thing and "scalar lists" are not included in that.  there is no "map of a scalar value" concept in SQLAlchemy and I dont even know how that would be possible


> 
> > you can only have collections of rows in other tables
> 
> The association table can exist without a foreign key to the other table. For the sake of example, let's say it's a User-Car many-to-many mapping (where users may own/use multiple cars, and cars may have multiple owners).
> 
> Now consider a "micro-service" (well, distributed) architecture, where the user identity (and details) exists in a authentication/identity system and the car data exists another system. In this case, relationship to the user is managed with a simple "user_id" column (not a foreign key) and any user details are fetched on-the-fly, where needed. This is the scenario I'm mapping (or rather, converting an old monolithic system/DB into).

there's a table and each row in the table indicates an element in the collection.   there is no DDL for a "scalar value" construct in SQL.  it's rows.   the ORM maps to rows.    associationproxy bridges the difference to have collections of scalars.  you seem to be arguing, "we shouldnt need the association proxy to do that".   So we're at where me and other SQLAlchemy devs have worked hard to make sure a certain use case is possible and I guess you're arguing that it's not good enough, OK!   sorry I cant help

**Reply by mtvx** [reply]
I'm definitively not arguing that something is wrong / bad, I'm sorry if I gave that impression. It was absolutely not the intent.

I tried to describe a common use case, that I thought could be maybe supported further with some extra constructs - should be nothing bad in community (politely) suggesting new features. And indeed, sometimes the community misunderstands and might ask/propose stupid things.

I'm obviously failing to understand some part of the (relational DB) context here. That happens.

Anyway, I assume the answer to the question, if the other-table-does-not-exist use case can be supported with `Table` association, is no, one has to use association object instead. That's fine.

Thanks for your persistent work on SQLAlchemy, it's awesome.

**Reply by zzzeek** [reply]
there's just no reason to add an additional layer that will see `Mapped[list[int]]` and then make a bunch of guesses that the association proxy is actually what's wanted, with a certain table name, a certain structure of that table, etc.  this is all rabbit hole stuff.  the pattern you want is possible, but SQLAlchemy likes to keep APIs close to the metal and explicit as this greatly reduces surprises for users as well as our own workload.    This is the opposite of the Django-like mindset that is all about one-line-of-code == magic.   we're the opposite of that as much as we can be.

**Reply by mtvx** [reply]
Well, I don't disagree with any of that. Maybe the word "magic" that I used in the original message was a mistake - what I meant was a "well-defined abstraction", or such, but chose a _cooler_ word. Anyway...
